src/cpp/session/modules/SessionEnvironment.R

```diff
@@ -364,16 +377,83 @@
 
 .rs.addFunction("functionNameFromCall", function(val)
 {
-   call <- attr(val, "_rs_call")
+   call <- .rs.nullCoalesce(
+      attr(val, "_rs_call", exact = TRUE),
+      val
+   )
+   
    if (is.function(call[[1]]))
-      "[Anonymous function]"
+      return("[Anonymous function]")
+   
+   if (is.name(call[[1L]]))
+      return(as.character(call[[1L]]))
+   
+   .rs.deparse(call[[1L]])
+})
+
+.rs.addFunction("sanitizeCallSummary", function(object)
+{
+   if (is.call(object))
+   {
+      # if this is the concatenation of a large number of objects,
+      # then just use a shorter representation of the call
+      callee <- object[[1L]]
+      long <-
+         is.name(callee) &&
+         length(object) > 20
+      
+      if (long)
+         return(as.call(list(callee, quote(...))))
+      
+      # sanitize each call entry separately
+      for (i in seq_along(object))
+         object[[i]] <- .rs.sanitizeCallSummary(object[[i]])
+      
+      # return object
+      object
+   }
+   else if (!is.language(object))
+   {
+      # if the object would be very expensive to deparse,
+      # just use a placeholder instead
+      #
+      # note that we still want to accept literals here,
+      # e.g. 42, "abc"
+      if (!is.object(object) && length(object) == 1)
+      {
+         object
+      }
+      else
+      {
+         type <- .rs.explorer.objectType(object)
+         as.name(sprintf("<%s>", type))
+      }
+   }
    else
-      as.character(substitute(call))
```

Comment:
Note: I believe this worked somewhat by accident, since calling `as.character()` on a call returns a character vector of the same length as the call. For example:

```
> call <- quote(c(a, b, c))
> as.character(call)
[1] "c" "a" "b" "c"
```

But since we only ever looked at the first element of the vector, it just so happened to produce the right result.

https://github.com/rstudio/rstudio/blob/58a3c0f1d2a51977f9d5e24b956ab02f948a0fcf/src/cpp/r/RSexp.cpp#L884-L894

(note that we just extract the first element of the associated vector, and silently ignore the rest)